### DISCUSSION TITLE
Change behaviour of Attribute values without quotes

### DISCUSSION THREAD

**Discussion by wrabit** [wrabit/django-cotton#141] [Ideas]
I've noticed people reach sub-conciously for the pattern that exists in django whereby if you want to provide template variables or native types as a attributes, They do so without quotes.

```html
attr=1
attr=variable2
```

For passing these, cotton make's its users following the dynamic attribute pattern like:

```
:attr="1" <!-- will be passed as :int = 1 -->
:attr="variable2" <!-- will be passed as template variable with that name -->
```

When you try and provide the values without quotes in cotton, quotes will be automatically applied and they will be parsed as strings.

Making a note because I think to support the non quoted way.

**Comment by mattbha** [comment]
My opinion

assume **views.py** passes (to template render) `context = {"myintvar": 1, "mystrvar": "hello"}`

```
:abc=1   # should be passed as int 1
:abc="1"  # should be passed as a str 1
:abc=myintvar  # should forward the context variable exactly as-is (as an int)
:abc="{{ myintvar }}"  # should render as a str, and pass the string
:abc=mystrvar  # should forward the context variable exactly as-is (as str)
abc=mystrvar  # should raise a syntax error? or be auto quoted
abc="{{ mystrvar }}"  # should render fine, as a str
abc=myintvar  # should render fine, as an int; or perhaps be autoquoted? (to make consistent)

```

in the case of
```
:attr="myintvar"
```

I think that should be passing a static string literal "myintvar" and not be automatically taking the variable value

IE keep
`:xyz` as a consistant way to pass forward variables, unique to django-cotton
`xyz` as a standard as-expected part of the render format, would work without django-cotton


open to other opinions

**Comment by wrabit** [comment]
When I built the first version of cotton, with the mindset of bringing html-like syntax, I was put off omitting `"` on attribute values, hence the `:` usage. I feel it serves the purpose quite well, but as time has passed, I too have wondered why we can't align further with native django template and allow quoteless values to be parsed directly as variables or string literals. The only situation where this won't work is where the value has spaces (i.e. a dict, list, tuple etc) because we wont have quotes to delimit the value... so choosing the `:` approach for these occasions should be fine, (people are free to use the ":" throughout in any case)

To overlay my thoughts on the examples you provided and to group by what is currently supported/documented and what might be:

## Quoted 

```python
:abc="1"  # int
:abc="myintvar"  # int
:abc="{{ myintvar }}" # this doesn't really have much use and is more efficient as :abc="myintvar"
:abc="[{{ myintvar }}]" # relating to the previous, an example that would be parsed as a list `[1]`
abc="{{ mystrvar }}"  # str
```

## Quoteless idea

```python
:abc=1   # int
:abc=myintvar  # as per variable - int <!-- maybe this will be the syntax error, as mixing ":" with quoteless -->
:abc=mystrvar  # as per variable - str <!-- same as above --> 
abc=mystrvar  # this is the difference about this discussion, I am considering it be the same as :abc="mystrvar" (str)
abc=myintvar  # again, as above, same as :abc="myintvar" (int)
```

**Comment by mattbha** [comment]
2 points

1) High level I personally think quotes-or-not should be a way to differentiate passing a context variable or a static value to the component. i feel strongly if its quoted, it should be static (matches what django does ie https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#extends also see bottom)

as in `:abc="myintvar"`) - i think this should be passing a static value, as indicated by the quotes
`:abc=myintvar` - i think this should be forwarding the (python typed) variable to the component


2) i dont feel strongly on this but I think `:key=` vs `key=` should be a way of indicating "special cotton python type" vs more traditional html attribute, which might require quoting (unless its a variable)

im not sure there would be a difference in `abc=myintvar` vs `:abc=myintvar` then, tho



I think it makes sense to match and extend the default django builtins with regards to includes and subtemplates. valid django syntax is:

`{% include "name_snippet.html" with greeting=greeting_variable only %}`

or

`{% include "name_snippet.html" with greeting="hello world" only %}`

where `only` can be removed to send full context, but ive never used this

(from https://docs.djangoproject.com/en/5.1/ref/templates/builtins/#include)

**Reply by wrabit** [reply]
I can't think of an example why I would want to reach for `:abc="myintvar"` to return the static value (string) of the evaluated variable. Besides that I think we're pretty much on the same path.

**Reply by mattbha** [reply]
I agree the use case is probably non-existent; I think it's less of a "is this useful" and more of a "following convention that variables aren't quoted".

**Reply by mattbha** [reply]
> There should be one-- and preferably only one --obvious way to do it.

If we allow both it it could become less obvious the "right" way

quoted value = static
nonquoted value = variable


from what i understand this would match django

unless there is something else or a usecase i am missing. i started a new project to purely use django-cotton as a poc and its been working great with this method
